### PR TITLE
Add exception to overseer communication and write up local and remote disputes

### PR DESCRIPTION
**Pull request by rphmeier**

### CODE REVIEW

File: roadmap/implementors-guide/guide.md [hunk 1]
```diff
@@ -883,8 +883,67 @@ Included: Option<()>,
 
 ### The Validity Module
 
+After a backed candidate is made available, it is included and proceeds into an acceptance period during which validators are randomly selected to do secondary checks of the parablock. Any reports disputing the validity of the candidate will cause escalation, where even more validators are requested to check the block, and so on, until either the parablock is determined to be invalid or valid. Those on the wrong side of the dispute are slashed and, if the parablock is deemed invalid, the relay chain is rolled back to a point before that block was included.
```
Comment (burdges) [inline roadmap/implementors-guide/guide.md:886]: I write say "(secondary) approval checks" or similar, but avoid saying just "secondary checks" since someone might read secondary as less important.

File: roadmap/implementors-guide/guide.md [hunk 2]
```diff
@@ -883,8 +883,67 @@ Included: Option<()>,
...
 
+One first question is to ask why different logic for local disputes is necessary. It seems that local disputes are necessary in order to create the first escalation that leads to block producers abandoning the chain and making remote disputes possible.
+
+Local disputes are only allowed on parablocks that have been included on the local chain and are in the acceptance period.
+
+For each such parablock, it is guaranteed by the inclusion pipeline that the parablock is available and the relevant validation code is available.
+
+Disputes may occur against blocks that have happened in the session prior to the current one, from the perspective of the chain. In this case, the prior validator set is responsible for handling the dispute and to do so with their keys from the last session. This means that validator duty actually extends 1 session beyond leaving the validator set.
```
Comment (burdges) [inline roadmap/implementors-guide/guide.md:911]: Yes, the VRF keys used should be whatever keys existed when the candidate became available.

File: roadmap/implementors-guide/guide.md [hunk 3]
```diff
@@ -883,8 +883,67 @@ Included: Option<()>,
...
+When a dispute has occurred on another fork, we need to transplant that dispute to every other fork. This poses some major challenges.
+
+There are two types of remote disputes. The first is a remote roll-up of a concluded dispute. These are simply all attestations for the block, those against it, and the result of all secondary checks. A concluded remote dispute can be resolved in a single transaction as it is an open-and-shut case of a quorum of validators disagreeing with another.
+
+The second type of remote dispute is the unconcluded dispute. An unconcluded remote dispute is started by any validator, using these things:
+  - A candidate
+  - The session that the candidate has appeared in.
+  - Backing for that candidate
+  - The validation code necessary for validation of the candidate. [TODO: optimize by excluding in case where code appears in `Paras::CurrentCode` of this fork of relay-chain]
+  - Secondary checks already done on that candidate, containing one or more disputes by validators. None of the disputes are required to have appeared on other chains. [TODO: validator-dispute could be instead replaced by a fisherman w/ bond]
```
Comment (burdges) [inline roadmap/implementors-guide/guide.md:926]: We could construct remote disputes as a series of related transactions, so individually these component transactions may not suffice, until we assemble all the component transactions.

File: roadmap/implementors-guide/guide.md [hunk 4]
```diff
@@ -883,8 +883,67 @@ Included: Option<()>,
...
+
+When beginning a remote dispute, at least one escalation by a validator is required, but this validator may be malicious and desires to be slashed. There is no guarantee that the para is registered on this fork of the relay chain or that the para was considered available on any fork of the relay chain.
+
+So the first step is to have the remote dispute proceed through an availability process similar to the one in [the Inclusion Module](#The-Inclusion-Module), but without worrying about core assignments or compactness in bitfields.
```
Comment (burdges) [inline roadmap/implementors-guide/guide.md:930]: I'm think a second availability game almost never helps because the pre-dispute availability game requires adversarial nodes vote available.  We could run an availability game with a lower threshold `f+1` which then says the block is reconstructable, unless those nodes lie.  

We could just say that one dispute authorizes gossiping authenticated pieces, which works both local and remote.  We pay for the dispute by extracting some fee from the disputer, which they massively recoup when they win the dispute, and governance refunds if strange shit happens.  

We'd need some fee even to run an availability game, so making the fee a bit larger justifies gossiping all the pieces, and simplifies the code significantly.

File: roadmap/implementors-guide/guide.md [hunk 5]
```diff
@@ -883,8 +883,67 @@ Included: Option<()>,
...
+
+We assume that remote disputes are with respect to the same validator set as on the current fork, as BABE and GRANDPA assure that forks are never long enough to diverge in validator set [TODO: this is at least directionally correct. handling disputes on other validator sets seems useless anyway as they wouldn't be bonded.]
+
+As with local disputes, the validators of the session the candidate was included on another chain are responsible for resolving the dispute and determining availability of the candidate.
```
Comment (burdges) [inline roadmap/implementors-guide/guide.md:934]: It's slightly hard to parse this sentence, but it seemingly says the disputed chain's validator set resolves the dispute on the good chain?  I'm not so sure..

File: roadmap/implementors-guide/guide.md (hunk 4 above)
Comment (AlistairStewart) [reply roadmap/implementors-guide/guide.md:930]: We would only need a new process if there wasn't already one on this chain or it was very old.

After some discussion with Jeff, we decided that a proof that availability was decided on the old chain shouldn't be too hard, provided we fix timeout logic in such a way that we can understand it. Any votes for a particular bitfield, after the backing transaction was included and before the timeout are either for this block or else we did decide that it was avilable before they could be votes for a future block. This is particularly sensible if we still have the backing transaction in the new fork.

The parathread like process is the only way we can handle older disputes, especially if we start dealing with some information that we need being pruned or session changes. But it sounds harder to code.

Comment (AlistairStewart) [reply roadmap/implementors-guide/guide.md:930]: We can just gossip the block. As long as we have faith that this works, we can slash evryone who attested to the block a large amount if it doesn't.

Comment (burdges) [reply roadmap/implementors-guide/guide.md:930]: I'm seemingly winning @AlistairStewart over to my view, but with backers paying, not the disputer.  We probably want both backers and disputer to pay I think.

File: roadmap/implementors-guide/guide.md (hunk 2 above)
Comment (rphmeier) [reply roadmap/implementors-guide/guide.md:911]: There is no mention of VRFs in this paragraph. But if you are referring to the remote disputes section, then I actually disagree with you, because there may be multiple forks where the candidate became available, so there is no clear definition of your statement.

File: roadmap/implementors-guide/guide.md (hunk 3 above)
Comment (rphmeier) [reply roadmap/implementors-guide/guide.md:926]: Right - 1 transaction to open the dispute, periodic transactions to update the dispute, and then after closing, accept transactions to add evidence to the losing side.

File: roadmap/implementors-guide/guide.md (hunk 5 above)
Comment (rphmeier) [reply roadmap/implementors-guide/guide.md:934]: Yes, and because we buffer validator sets they should be the same. More specifically, a block included in session X on the bad chain should use the validators of session X on the good chain, not the validators of session X+k. But the validators of session X would be the same on both chains.

If they aren't, then there's no point even attempting to slash a remote session X, as they won't be guaranteed to be bonded.

File: roadmap/implementors-guide/guide.md (hunk 4 above)
Comment (rphmeier) [reply roadmap/implementors-guide/guide.md:930]: The availability proof actually is not easy and interferes quite a lot with state pruning.

**Comment by rphmeier** [comment]
I've also been thinking a bit about what happens when a chain is reverted. In practice, I think we want something like this:
  1. There is a special inherent that can be included in a block which removes a candidate from pending availability and contains all the information to start a remote dispute about that candidate, including an escalation (typically by the block author)
  2. When reverting a chain or realizing that an escalation will lead to eventual reversion of the chain, as a validator we can include that inherent in any blocks we author on chains where that candidate has not been disputed.
  3. This will have the side effect of removing the candidate from availability on any chains where it is backed

**Comment by burdges** [comment]
I think our scheme should go roughly: 

We've the availability distribution design that distributes pieces for candidates backed in the current chain directly using QUIC.  We've an execution intrinsic(?) for when a candidate actually runs that checks the availability bitfields. 

We've a dispute extrinsic(?) that consists of a candidate receipt with at least one backer, but not necessarily enough backing, and a signature by a disputer.  A dispute extrinsic can be posted to any chain with several consequences.
- Availability: We gossip the candidate's availability pieces among all validators.  Any validator who receives 1/3rd shall check the block and post their vote.  We do not halt normal availability distribution for the candidate.  
- Votes:  
  - Votes cannot be gossiped or posted without the dispute extrinsic, but otherwise they can be copied to any chain, like the dispute extrinsic itself.  
  - If `2f+1` vote valid or invalid then the loosing side gets slashed 100%, and non-voters eventually have their earnings penalized but not booted. 
  - If neither side gets `2f+1` then the candidate is deemed invalid, but neither side gets slashed per se, and instead all valid voters including backers have their earnings reduced to compensate for the distraction but again not booted.  
- Fork choice:
  - If the associated execution intrinsic was not yet posted, then it cannot be posted unless the vote concludes the candidate is valid.  We could maybe weaken this to relay chain block producers who vote valid being allowed to post it, but probably unhelpful. 
  - If the associated execution intrinsic was posted, then the chain becomes invalid once we reach `f+1` invalid votes.  We need not abandon chain with less than `f+1` invalid votes, but our fork choice rule could anticipate the vote in various way.  Any relay chain block producer who votes invalid cannot build upon chains containing the associated execution intrinsic.

File: roadmap/implementors-guide/guide.md (hunk 4 above)
Comment (burdges) [reply roadmap/implementors-guide/guide.md:930]: I posted a suggestion below.

**Comment by rphmeier** [comment]
Generally looks good.

> instead all valid voters including backers have their earnings reduced to compensate for the distraction but again not booted

Not sure how we do this. @thiolliere is there a reward points mechanism for this? @burdges over what period do we need to reduce earnings? @alfonsocev how much do we need to do this, economically speaking?

> If the associated execution intrinsic was not yet posted, then it cannot be posted unless the vote concludes the candidate is valid. We could maybe weaken this to relay chain block producers who vote valid being allowed to post it, but probably unhelpful.

If the candidate is in the availability stage and occupying an availability core, I'd rather just remove the thing from availability so that it doesn't occupy an availability core the entire time the dispute is ongoing. If the candidate turns out to be good, the validators will get paid for their time, but the para could proceed regardless.

Not removing the thing from availability means we need to roll back further than strictly necessary.

> Any relay chain block producer who votes invalid cannot build upon chains containing the associated execution intrinsic.

This is hard to enforce. But worth noting that any relay chain fork that has included the invalid candidate will have a remote dispute started on it that will eventually blacklist the chain, if it has any chance of becoming a viable chain.

**Comment by burdges** [comment]
We should avoid creating any mechanism for booting validators who maybe did nothing wrong, so "reduce earnings" means "minor slash without booting".  We might require the slashing system here simply because the validators could come from some other chain.  We can support that more easily in the slashing system than in the earnings system.  We could otoh avoid the slashing system by restricting to the validators' self stake, assuming this stays cheap enough, which makes it simple across chains.

Yes, I suppose parathread want those cores liberated fast.  We've changed the availability scheme from directed to gossiped to everyone, so yes we could just free up the core here.  

We've lingering concerns about censoring parachains/parathreads here, or even censoring the relay chain with roll backs, which complicates saying when we fork the chain.  In principle, we often fork only a block or so, and continue with the availability game almost done, but yes the dispute replaces the availability game.

**Comment by burdges** [comment]
I'll clarify for others:  We cannot penalize anyone for not voting without an availability roll up, but doing a roll up sucks.  We're discussing the "no roll up" design that
1. penalizes non-voters only when one sides gets `2f+1`, and 
2. makes "valid" voters pay distraction fees when neither side gets `2f+1` votes. 

We need 1 to compensate for the harmful incentives created by 2 of course.

File: roadmap/implementors-guide/guide.md [hunk 6]
```diff
@@ -937,13 +937,16 @@ If the candidate was not made available on another fork of the relay chain, the
...
-  - use VRF from relay-chain block on another fork where the parablock was included. There are likely multiple choices for this.
-
 After enough validator self-select, under the same escalation rules as for local disputes, the Remote dispute will conclude, slashing all those on the wrong side of the dispute. After concluding, the remote dispute remains open for a set amount of blocks to accept any further proof of additional validators being on the wrong side.
 
+### Slashing and Incentivization
+
+The goal of the dispute is to garner a 2/3+ (2f + 1) supermajority either in favor of or against the candidate.
+
+For remote disputes, it is possible that the parablock disputed has never actually passed any availability process on any chain. In this case, validators will not be able to obtain the PoV of the parablock and there will be relatively few votes. We want to disincentivize voters claiming validity of the block from preventing it from becoming available, so we charge them a small distraction fee for wasting the others' time if the dispute does not garner a 2/3+ supermajority on either side. This fee can take the form of a small slash or a reduction in rewards.
```
Comment (burdges) [inline roadmap/implementors-guide/guide.md:946]: ".. a small slash that does not chill or .."

We should discuss if the distraction fee should chill/boot validators @AlistairStewart 

Pro-chill: Anyone who voted valid could gossip the pieces themselves. 
Anti-chill: Attackers could file false disputes so that DoS attacks cause chilling.  

Can DoS attack chill anyways via ImOnline?  If so, maybe fix that one day..

File: roadmap/implementors-guide/guide.md [hunk 7]
```diff
@@ -937,13 +937,16 @@ If the candidate was not made available on another fork of the relay chain, the
...
+
+When a supermajority is achieved for the dispute in either the valid or invalid direction, we will penalize non-voters either by issuing a small slash or reducing their rewards. We prevent censorship of the remaining validators by leaving the dispute open for some blocks after resolution in order to accept late votes.
```
Comment (burdges) [inline roadmap/implementors-guide/guide.md:948]: Again "that does not chill" probably

File: roadmap/implementors-guide/guide.md (hunk 6 above)
Comment (rphmeier) [reply roadmap/implementors-guide/guide.md:946]: Pro-chill: currently only nonzero slash chills.

Comment (rphmeier) [reply roadmap/implementors-guide/guide.md:946]: @burdges What do you say about deferring this to a follow-up PR? I don't believe it substantially alters the meaning of the text, but is worth addressing

Comment (burdges) [reply roadmap/implementors-guide/guide.md:946]: Sure
